File: frontend/src/types/engagement.ts
export type EngagementResourceType =
  | 'did_you_know'
  | 'prior_knowledge'
  | 'detonating_question'
  | 'real_news'
  | 'mini_quiz'
  | 'short_challenge'

export interface MiniQuizMetadata {
  options: string[]
  correct_index: number
  explanation: string
}

export interface ShortChallengeMetadata {
  prompt: string
  hint?: string
  expected_keywords?: string[]
}

export interface EngagementBadge {
  slug: string
  label: string
  icon: string
  xp: number
}

export interface EngagementResource {
  id: string
  resource_type: EngagementResourceType
  title: string
  content: string
  media_url?: string | null
  modality_target?: string | null
  display_order: number
  is_interactive: boolean
  resource_metadata: MiniQuizMetadata | ShortChallengeMetadata | Record<string, unknown>
}

export type EngagementSessionStatus = 'pending' | 'active' | 'completed' | 'skipped'

export interface EngagementSession {
  session_id: string
  status: EngagementSessionStatus
  modality_profile?: string | null
  resources: EngagementResource[]
  xp_earned: number
  resources_shown: number
  earned_badges: EngagementBadge[]
}

export type InteractionType = 'view' | 'answer' | 'submit' | 'skip'

export interface InteractRequest {
  session_id: string
  resource_id: string
  interaction_type: InteractionType
  time_spent_seconds?: number
  response_data?: Record<string, unknown>
}

export interface InteractResponse {
  ok: boolean
  xp_delta: number
  xp_total: number
  is_correct?: boolean | null
  badge?: EngagementBadge | null
}

export interface CompleteRequest {
  session_id: string
  skipped?: boolean
}

export interface CompleteResponse {
  xp_earned: number
  xp_breakdown: Record<string, number>
  badges: EngagementBadge[]
  next_step: string
  message: string
}

File: backend/app/schemas/engagement.py
"""
Schemas Pydantic para el sistema Engage.

Tres flujos:
  GET  /api/engagement/start         → EngagementSessionOut
  POST /api/engagement/interact      → InteractRequest / InteractResponse
  POST /api/engagement/complete      → CompleteRequest / CompleteResponse
"""

from __future__ import annotations
from typing import Any, Optional
from pydantic import BaseModel, Field


# ── Outbound ────────────────────────────────────────────────────────────────────

class EngagementResourceOut(BaseModel):
    id:               str
    resource_type:    str
    title:            str
    content:          str
    media_url:        Optional[str] = None
    modality_target:  Optional[str] = None
    display_order:    int
    is_interactive:   bool
    resource_metadata: dict[str, Any] = Field(default_factory=dict)

    model_config = {"from_attributes": True}


class BadgeOut(BaseModel):
    slug:  str
    label: str
    icon:  str
    xp:    int


class EngagementSessionOut(BaseModel):
    session_id:          str
    status:              str
    modality_profile:    Optional[str]
    resources:           list[EngagementResourceOut]
    xp_earned:           int
    resources_shown:     int
    earned_badges:       list[BadgeOut] = Field(default_factory=list)

    model_config = {"from_attributes": True}


class InteractResponse(BaseModel):
    ok:         bool
    xp_delta:   int
    xp_total:   int
    is_correct: Optional[bool] = None
    badge:      Optional[BadgeOut] = None   # insignia desbloqueada por esta interacción


class CompleteResponse(BaseModel):
    xp_earned:    int
    xp_breakdown: dict[str, int]
    badges:       list[BadgeOut]
    next_step:    str
    message:      str


# ── Inbound ─────────────────────────────────────────────────────────────────────

class InteractRequest(BaseModel):
    session_id:         str
    resource_id:        str
    interaction_type:   str = Field(
        ...,
        description="view | answer | submit | skip",
    )
    time_spent_seconds: int = Field(default=0, ge=0)
    response_data:      dict[str, Any] = Field(default_factory=dict)


class CompleteRequest(BaseModel):
    session_id: str
    skipped:    bool = False
